From 4919565c152f2ecee760b5419c947764e991ab3e
From: Cliff Frey <cliff@meraki.com>
Date: Tue, 07 Aug 2012 15:59:40 +0000
Subject: lineedit: fix Alt-D when cursor==0

Signed-off-by: Cliff Frey <cliff@meraki.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
From b2320370be14811459718b9fe418efed75ea3615
From: Denys Vlasenko <vda.linux@googlemail.com>
Date: Thu, 27 Sep 2012 14:03:49 +0000
Subject: lineedit: in !EDITING config, return -1 on fgets error

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
From 46031da862a60422f80050a905cea0b67026b021
From: Shawn J. Goff <shawn7400@gmail.com>
Date: Wed, 27 Feb 2013 17:30:05 +0000
Subject: lineedit: initialize delptr

In vi mode, the 'p' and 'P' commands caused a segfault when nothing had
been put in the buffer yet because the delptr was not initialized.

Signed-off-by: Shawn J. Goff <shawn7400@gmail.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---

diff --git a/libbb/lineedit.c b/libbb/lineedit.c
--- a/libbb/lineedit.c
+++ b/libbb/lineedit.c
@@ -187,6 +187,7 @@ extern struct lineedit_statics *const lineedit_ptr_to_statics;
 	cmdedit_termw = 80; \
 	IF_FEATURE_EDITING_FANCY_PROMPT(num_ok_lines = 1;) \
 	IF_USERNAME_OR_HOMEDIR(home_pwd_buf = (char*)null_str;) \
+	IF_FEATURE_EDITING_VI(delptr = delbuf;) \
 } while (0)
 
 static void deinit_S(void)
@@ -2527,9 +2527,9 @@ int FAST_FUNC read_line_input(line_input_t *st, const char *prompt, char *comman
 					/* Delete word forward */
 					int nc, sc = cursor;
 					ctrl_right();
-					nc = cursor;
-					input_backward(cursor - sc);
-					while (--nc >= cursor)
+					nc = cursor - sc;
+					input_backward(nc);
+					while (--nc >= 0)
 						input_delete(1);
 					break;
 				}
@@ -2729,7 +2729,8 @@ int FAST_FUNC read_line_input(const char* prompt, char* command, int maxsize)
 {
 	fputs(prompt, stdout);
 	fflush_all();
-	fgets(command, maxsize, stdin);
+	if (!fgets(command, maxsize, stdin))
+		return -1;
 	return strlen(command);
 }
 
--
cgit v0.9.0.1-2-gef13
